mlflow/azureml/__init__.py

```diff
@@ -206,8 +207,163 @@ def build_image(model_uri, workspace, image_name=None, model_name=None,
+    # The Azure ML SDK is only compatible with Python 3. However, the `mlflow.azureml` module should
+    # still be accessible for import from Python 2. Therefore, we will only import from the SDK
+    # upon method invocation.
+    # pylint: disable=import-error
+    from azureml.core.model import Model as AzureModel, InferenceConfig
+    from azureml.core import Environment as AzureEnvironment
+    from azureml.core import VERSION as AZUREML_VERSION
+
+    absolute_model_path = _download_artifact_from_uri(model_uri)
+
+    model_pyfunc_conf, run_id = _load_pyfunc_conf(model_path=absolute_model_path)
+    model_python_version = model_pyfunc_conf.get(pyfunc.PY_VERSION, None)
+    if model_python_version is not None and\
+            StrictVersion(model_python_version) < StrictVersion("3.0.0"):
+        raise MlflowException(
+                message=("Azure ML can only deploy models trained in Python 3 and above. See"
+                         " the following MLflow GitHub issue for a thorough explanation of this"
+                         " limitation and a workaround to enable support for deploying models"
+                         " trained in Python 2: https://github.com/mlflow/mlflow/issues/668"),
+                error_code=INVALID_PARAMETER_VALUE)
+
+    tags = _build_tags(model_uri=model_uri, model_python_version=model_python_version)
+
+    if service_name is None:
+        service_name = _get_mlflow_azure_name_from_run(run_id)
+    if model_name is None:
+        model_name = _get_mlflow_azure_name_from_run(run_id)
+
+    with TempDir(chdr=True) as tmp:
+        model_directory_path = tmp.path("model")
+        tmp_model_path = os.path.join(
+            model_directory_path,
+            _copy_file_or_tree(src=absolute_model_path, dst=model_directory_path))
+
+
+        registered_model = AzureModel.register(workspace=workspace, model_path=tmp_model_path,
+                                               model_name=model_name, tags=tags)
+
+        _logger.info("Registered an Azure Model with name: `%s` and version: `%s`",
+                     registered_model.name, registered_model.version)
+
+        # Create an execution script (entry point) for the image's model server. Azure ML requires
+        # the container's execution script to be located in the current working directory during
+        # image creation, so we create the execution script as a temporary file in the current
+        # working directory.
+        execution_script_path = tmp.path("execution_script.py")
+        _create_execution_script(output_path=execution_script_path, azure_model=registered_model)
+
+        environment = None
+        if pyfunc.ENV in model_pyfunc_conf:
+            environment = AzureEnvironment.from_conda_specification(_get_mlflow_azure_name_from_run(run_id), os.path.join(tmp_model_path, model_pyfunc_conf[pyfunc.ENV]))
+        else:
+            raise MlflowException("Cannot deploy pyfunc model without an environment to Azure ML.")
```

Comment:
@csteegz Why does this restriction exist? We don't currently perform validation for the presence of a conda environment in the `build_image` routine; in fact, the `conda_file` argument of `ContainerImage.image_configuration()` (which is called/constructed during `mlflow.azureml.build_image`) has a default value of `None`, which seems to imply that it's optional: https://docs.microsoft.com/en-us/python/api/azureml-core/azureml.core.image.containerimage?view=azure-ml-py#image-configuration-execution-script--runtime--conda-file-none--docker-file-none--schema-file-none--dependencies-none--enable-gpu-none--tags-none--properties-none--description-none--base-image-none--base-image-registry-none--cuda-version-none-

Can you provide more context here? Did we miss something in `build_image`, or is there a requirements difference between the Azure resources created by `build_image` and `deploy` regarding the `conda.yaml` file?

Reply:
the containerimage class is deprecated in favor of environments

Reply:
I think there is better defaulting we can do here that @dbczumar's question leads to - it's just a rejiggering of concepts but ContainerImage  implied a certain default Environment that we should add here as well

Reply:
@csteegz  is this thread actually resolved? can we do the equivalent of python/python-slim like the old APIs used to do here while perhaps emitting a warning?

Reply:
Uh - What I switched to do was use a default environment with azureml-defaults and MLFlow added, which should be the equivalent of python-slim. Not sure why you weren't able to see that change (https://github.com/mlflow/mlflow/pull/2375/commits/af7d320f22296b317040577e190a4388b1209a83#diff-6e19c0d7aba3a117d4d11e9b592639a5R346)